Src/FluentAssertions/Common/ExpressionExtensions.cs

```diff
@@ -87,29 +88,51 @@ public static MemberPath GetMemberPath<TDeclaringType, TPropertyType>(
                 case ExpressionType.Call:
                     var methodCallExpression = (MethodCallExpression)node;
 
-                    if (methodCallExpression is not { Method.Name: "get_Item", Arguments: [ConstantExpression argumentExpression] })
+                    if (methodCallExpression is not
+                        { Method.Name: "get_Item", Arguments: [ConstantExpression argumentExpression] })
                     {
                         throw new ArgumentException(GetUnsupportedExpressionMessage(expression.Body), nameof(expression));
                     }
 
                     node = methodCallExpression.Object;
-                    segments.Add($"[{argumentExpression.Value}]");
+                    singlePath = $"[{argumentExpression.Value}].{singlePath}";
+                    break;
+                case ExpressionType.New:
```

Comment:
♻️ This entire method is become too bloated and needs some refactoring (but maybe not now). An implementation of the Strategy Pattern might be a better fit.